### ISSUE TITLE
Suggestion: add keybindings for vdiff

### ISSUE THREAD

**Issue by edkolev** [jojojames/evil-collection#70]
Hi, and thanks for this package!

I would like to suggest adding the following [vdiff](https://github.com/justbur/emacs-vdiff) keybindings.

```elisp
  (evil-define-minor-mode-key 'normal 'vdiff-mode
    "]c" 'vdiff-next-hunk
    "[c" 'vdiff-previous-hunk
    "go" 'vdiff-receive-changes
    "gp" 'vdiff-send-changes)
```

The `go` and `gp` are intended to mimic vim's `do` and `dp`. I'm not aware of a sane universal way to bind `do` and `dp` without causing possible breakages. The best option I know is provided by general.el, but I don't think it can be used in evil-collection.

I'll submit a PR if you think this is a good addition.

**Comment by noctuid** [comment]
I think the best approach here to mimic `do` and `dp` here would probably be to bind `o` and `p` in operator state, which is how some packages like evil-surround do it. That works for evil-surround since it can be used with any operator, but in this case it should only work for `d`. Advising `evil-delete` or binding `d` to a command that temporarily bound `o` and `p` in `evil-operator-shortcut-map` would probably work. Something similar to this:

```elisp
(evil-define-operator evil-collection-vdiff-delete
  (beg end type register yank-handler)
  "..."
  (interactive "<R><x><y>")
  (define-key evil-operator-shortcut-map "o" #'vdiff-receive-changes)
  (define-key evil-operator-shortcut-map "p" #'vdiff-receive-changes)
  (evil-delete beg end type register yank-handler)
  (define-key evil-operator-shortcut-map "o" nil)
  (define-key evil-operator-shortcut-map "p" nil))
```

I tried using `let` so that `evil-operator-shortcut-map` is only temporarily altered instead of explicitly unbinding the keys, but that doesn't seem to work.

**Comment by edkolev** [comment]
Some time ago I played around with this. Here s the final semi-working snippet 
https://github.com/justbur/emacs-vdiff/issues/14#issuecomment-283597316

I'll try to come up with a solid implementation.

**Comment by noctuid** [comment]
I think using a menu item is preferable since the user can still bind `o`/`p` by default/for different operators.

**Comment by jojojames** [comment]
I'm not familiar with vimdiff but this seems like a good addition to have.

```
    "]c" 'vdiff-next-hunk
    "[c" 'vdiff-previous-hunk
```

Is this becaues vimdiff is using a buffer where ']' and '[' are used? Probably unrelated but reminds me of #64.

**Comment by Somelauw** [comment]
I don't know how menu-item works. Naively, I would write something like this:

```emacs-lisp
  (defvar evil-inhibited-o nil)
  (defvar evil-inhibited-p nil)

  (defun evil-inhibited-o ()
    (interactive)
    (when-let ((command (alist-get evil-this-operator evil-inhibited-o)))
      (setq evil-inhibit-operator t)
      (call-interactively command)))

  (defun evil-inhibited-p ()
    (interactive)
    (when-let ((command (alist-get evil-this-operator evil-inhibited-p)))
      (setq evil-inhibit-operator t)
      (call-interactively command)))

  (define-key evil-operator-state-map "p" 'evil-inhibited-p)
  (define-key evil-operator-state-map "o" 'evil-inhibited-o)

  ;; Now it can be used like this
  (setf (alist-get #'evil-delete evil-inhibited-o) #'vdiff-receive-changes)
  (setf (alist-get #'lispyville-delete evil-inhibited-o) #'vdiff-receive-changes)

  (setf (alist-get #'evil-delete evil-inhibited-p) #'vdiff-send-changes)
  (setf (alist-get #'lispyville-delete evil-inhibited-p) #'vdiff-send-changes)
```

I agree that it's useful to have this:
- defined in operator-state, so it works orthogonal to custom deletion commands such as lispyville-delete.
- extensible so that o/p can be defined differently for different operators

**Comment by justbur** [comment]
I’d like to have a more robust solution to binding keys like this in the evil repo. There are problems with how evil-surround does it (e.g., it has to hack together a way for repeat to work that is buggy) so mimicking evil-surround is not a great idea. 

You should be aware that to work properly repeat requires `pre-command-hook` `post-command-hook` to run with `this-command` set properly  among other things. 

I might take a shot at this if I find time.

**Comment by noctuid** [comment]
Repeat works without changes for the solution I posted.

**Comment by justbur** [comment]
@noctuid Making a new delete function to read the keys and dispatch to `evil-delete` will work, but it doesn't seem easy to generalize if you have two minor modes which want to alter the behavior of `evil-delete`. 

I think your proposal is the cleanest and avoids the problems that evil-surround runs into. I was hoping there was a nicer way that didn't mean that packages had to be aware of what other packages were doing. 

For an example you're familiar with, suppose you wanted to have lispyville enabled with your adjustment to `evil-delete` above? How do you imagine handling that?

I might be misunderstanding what you're suggesting though

**Comment by noctuid** [comment]
If you were talking about the first `evil-collection-vdiff-delete` example I gave, I was suggesting the menu-item example I gave after that, which will work regardless of what operator `d` is bound to as long as its in `evil-collection-delete-operators` (or some equivalent variable). If these packages all bound `[remap evil-delete]` instead of keys directly, it wouldn't be necessary to manually keep track of alternate delete commands.

If you were talking about `general-key-dispatch`, it would be possible to handle a situation where `[remap ...]` was used to change a command automatically. Otherwise, the user would have to intentionally hande it. I'm not suggesting using something like `general-key-dispatch` be used in packages though. Something like `general-key-dispatch` is more useful if the user wants to do this for a non-operator (e.g. bind `jk` in insert state or bind `q;` as the equivalent of `q:` without altering `evil-record-macro`) or wants an equivalent of vim's timeout functionality. Binding in operator state is probably the better way to go since it works for most cases where this sort of functionality is desirable in a package and is a lot simpler.
